File: parkingScript.js
require('dotenv').config();
const axios = require('axios');
process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

//Token 
async function getToken() {
    const url = `https://data.apk2.es/api/login`;
    try {
        const response = await axios.post(url, {
            username: process.env.username,
            secret: process.env.secret,
        });
        return response.data.token;
    } catch (error) {
        if (error.response) {
            console.error("Error en la respuesta:", error.response.data);
        } else if (error.request) {
            console.log ("Error en la petición:", error.request);
        } else {
            console.error("Error al realizar la petición:", error.message);
        }
        throw error;
    }
}

const fechaActual = new Date();
const fechaAnterior =  5*60*1000;
const fecha = new Date(fechaActual.getTime() - fechaAnterior); 

//Petición http(get) a la API
 async function getOcupation(){
    let cecos = "B149;B156;B157";
    let from = fecha.toISOString().slice(0, 19).replace('T', ' '); 
    let to = new Date().toISOString().slice(0, 19).replace('T', ' ');
    let nextToken = await getToken(); 
    const url = `https://data.apk2.es/api/parking-occupation/?cecos=${cecos}&from=${from}&to=${to}`;

    try {
        const response = await axios.get(url, {
            headers: {
                Authorization: `Bearer ${nextToken}`,
            },
    });
    return response.data;
    } catch (error) {
        if (error.response) {
            console.error("Error en la respuesta:", error.response.data);
        } else if (error.request) {
            console.log ("Error en la petición:", error.request);
        } else {
            console.error("Error al realizar la petición:", error.message);
        }
        throw error;
    }
} 
 
const plataforma = "smart.albacete.es";
const dispositivos = [
    {
      accessToken: process.env.tokenB156,
      id: "B156",
      totales: 749,
      abonados: 0,
      rotacional: 0,
      libres: 0,
      ocupado: 0,
      ocupacion: 0,
    },
    {
      accessToken: process.env.tokenB157,
      id: "B157",
      totales: 298,
      abonados: 0,
      rotacional: 0,
      libres: 0,
      ocupado: 0,
      ocupacion: 0,
    },
      {
      accessToken: process.env.tokenB156,
      id: "B149",
      totales: 225,
      abonados: 0,
      rotacional: 0, 
      libres: 0,
      ocupado: 0,
      ocupacion: 0,
      }
  ]; 

  async function sendAttributes(dispositivo) {
    const attributes = {
        "totales": dispositivo.totales,
    };

    try {
        await axios.post(`https://${plataforma}/api/v1/${dispositivo.accessToken}/attributes`, attributes, {
            headers: {
                'accept': 'application/json',
                'Content-Type': 'application/json'
            }
        });
    } catch (error) {
        if (error.response) {
            console.error("Error en la respuesta:", error.response.data);
        } else if (error.request) {
            console.log("Error en la petición:", error.request);
        } else {
            console.error("Error al realizar la petición:", error.message);
        }
        throw error;
    }
}

//Petición http(post) a la plataforma
async function sendTelemetry(dispositivo) {
    try { 
        const data = await getOcupation();
        const parkingDataMap = {};
        
              
        data.data.forEach(item => { 
            if (!parkingDataMap[item.parking] || new Date(item.time) > new Date(parkingDataMap[item.parking].time)) { 
                parkingDataMap[item.parking] = item; 
            }
        });

         const parkingData = parkingDataMap[dispositivo.id];
         
         if (parkingData) {
            dispositivo.abonados = parkingData.subscriber; 
            dispositivo.rotacional = parkingData.rotation; 
            dispositivo.id = parkingData.parking; 
        }
        dispositivo.libres = dispositivo.totales - dispositivo.abonados - dispositivo.rotacional;
        dispositivo.ocupado = dispositivo.totales - dispositivo.libres;
        dispositivo.ocupacion = dispositivo.ocupado * 100 / dispositivo.totales;
    
        const telemetryData = {

        abonados: dispositivo.abonados,
        rotacional: dispositivo.rotacional,
        libres: dispositivo.libres,
        ocupado: dispositivo.ocupado,
        ocupacion: dispositivo.ocupacion,
        };

        await axios.post(`https://${plataforma}/api/v1/${dispositivo.accessToken}/telemetry`, telemetryData,{
                headers: {
                    'accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            });
        console.log(`Datos de ${dispositivo.id} enviados correctamente`);
    } catch (error) {
        if (error.response) {
            console.error("Error en la respuesta:", error.response.data);
        } else if (error.request) {
            console.log ("Error en la petición:", error.request);
        } else {
            console.error("Error al realizar la petición:", error.message);
        }
        throw error;
    }
}

dispositivos.forEach(async dispositivo => {
    await sendTelemetry(dispositivo);
    await sendAttributes(dispositivo);
    
    setInterval(async() => { 
        try {
            await sendAttributes(dispositivo);
            await sendTelemetry(dispositivo);
        } catch (error) {
            console.error(`Error al cargar los datos de ${dispositivo.nombre}:`, error.message);
        }
    }, 300000);
}
);

File: parkingScript.py
import os
import requests
import json
import datetime
import time

# Configura el archivo .env
from dotenv import load_dotenv
load_dotenv()

# Token
def get_token():
    try:
        with open('tokensi.txt', 'r') as file:
            data = file.read()
            
        os.remove('tokensi.txt')
        return data
    except Exception as e:
        print(e)
    
    url = "https://data.apk2.es/api/login"
    try:
        response = requests.post(url, json={
            "username": os.getenv("username"),
            "secret": os.getenv("secret")
        })
        response_data = response.json()
        return response_data["token"]
    except requests.exceptions.RequestException as error:
        if hasattr(error, "response") and error.response:
            print("Error en la respuesta del Token:", error.response.json())
        else:
            print("Error al realizar la petición:", error)
        raise error

current_date = datetime.datetime.now()
delta = datetime.timedelta(hours=2)
current_date = current_date - delta
previous_date = current_date - datetime.timedelta(minutes=5)


# Petición HTTP (GET) a la API
def get_occupation():
    cecos = "B149;B156;B157"
    from_time = previous_date.strftime("%Y-%m-%d %H:%M:%S")
    to_time = current_date.strftime("%Y-%m-%d %H:%M:%S")
   
    next_token = get_token()
    
    url = f"https://data.apk2.es/api/parking-occupation/?cecos={cecos}&from={from_time}&to={to_time}"
    
    try:
        headers = {
            "Authorization": f"Bearer {next_token}"
        }
        
        response = requests.get(url, headers=headers)
        response_data = response.json()
        

        if response_data.get("nextToken") is not None:
            with open("tokensi.txt", "w") as file:
                file.write(response_data["nextToken"])

        print("Datos obtenidos de la API:", response_data)
        return response_data
    except requests.exceptions.RequestException as error:
        if hasattr(error, "response") and error.response:
            print("Error en la respuesta de la API:", error.response.json())
        else:
            print("Error al realizar la petición:", error)
        raise error

platform = "plataforma:9090"

devices = [
    {
        "accessToken": os.getenv("tokenB156"),
        "id": "B156",
        "totales": 749,
        "abonados": 0,
        "rotacional": 0,
        "libres": 0,
        "ocupado": 0,
        "ocupacion": 0
    },
    {
        "accessToken": os.getenv("tokenB157"),
        "id": "B157",
        "totales": 298,
        "abonados": 0,
        "rotacional": 0,
        "libres": 0,
        "ocupado": 0,
        "ocupacion": 0
    },
    {
        "accessToken": os.getenv("tokenB149"),
        "id": "B149",
        "totales": 225,
        "abonados": 0,
        "rotacional": 0,
        "libres": 0,
        "ocupado": 0,
        "ocupacion": 0
    }
]

# Petición HTTP (POST) a la plataforma
def send_attributes(dispositivo):
    attributes = {
        "totales": dispositivo["totales"]
    }
    try:
        headers = {
            "accept": "application/json",
            "Content-Type": "application/json"
        }
        url = f"https://{platform}/api/v1/{dispositivo['accessToken']}/attributes"
        response = requests.post(url, json=attributes, headers=headers, verify=False)
        response.raise_for_status()
    except requests.exceptions.RequestException as error:
        if hasattr(error, "response") and error.response:
            print("Error en la respuesta de los Atributos:", error.response.json())
        else:
            print("Error al realizar la petición:", error)
        raise error

# Petición HTTP (POST) a la plataforma
def send_telemetry(dispositivo):
    try:
        data = get_occupation()
        parking_data_map = {}
        
        for item in data["data"]:
            if item["parking"] not in parking_data_map or datetime.datetime.fromisoformat(item["time"]).strftime("%Y-%m-%d %H:%M:%S") > datetime.datetime.fromisoformat(parking_data_map[item["parking"]]["time"]).strftime("%Y-%m-%d %H:%M:%S"):
                parking_data_map[item["parking"]] = item

        parking_data = parking_data_map[dispositivo["id"]]

        if parking_data:
            dispositivo["abonados"] = parking_data["subscriber"]
            dispositivo["rotacional"] = parking_data["rotation"]
            dispositivo["id"] = parking_data["parking"]

        dispositivo["libres"] = dispositivo["totales"] - dispositivo["abonados"] - dispositivo["rotacional"]
        dispositivo["ocupado"] = dispositivo["totales"] - dispositivo["libres"]
        dispositivo["ocupacion"] = dispositivo["ocupado"] * 100 / dispositivo["totales"]

        telemetry_data = {
            "abonados": dispositivo["abonados"],
            "rotacional": dispositivo["rotacional"],
            "libres": dispositivo["libres"],
            "ocupado": dispositivo["ocupado"],
            "ocupacion": dispositivo["ocupacion"]
        }
        print("Datos enviados a la plataforma:", telemetry_data)

        headers = {
            "accept": "application/json",
            "Content-Type": "application/json"
        }
        url = f"https://{platform}/api/v1/{dispositivo['accessToken']}/telemetry"
        response = requests.post(url, json=telemetry_data, headers=headers, verify=False)
        response.raise_for_status()
        print(f"Datos de {dispositivo['id']} enviados correctamente")
    except requests.exceptions.RequestException as error:
        if hasattr(error, "response") and error.response:
            print("Error en la respuesta del sendTelemetryData:", error.response.json())
        else:
            print("Error al realizar la petición:", error)
        raise error

# Ejecutar las solicitudes para cada dispositivo y configurar un intervalo de 300 segundos (5 minutos)

while True:
    for dispositivo in devices:
        try:
            send_attributes(dispositivo)
            send_telemetry(dispositivo)
              # Esperar 5 minutos
        except Exception as error:
            print(f"Error al cargar los datos de {dispositivo['id']}:", error)

    time.sleep(300)